### PR TITLE
Virtualize Media Player Grid

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

Using Lit Virtualizer, make the media player more efficient and less taxing on older devices and mobile apps

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/media-player/ha-media-player-browse.ts [hunk 1]
```diff
@@ -147,6 +147,144 @@ export class HaMediaPlayerBrowse extends LitElement {
     }
   }
 
+  public willUpdate(changedProps: PropertyValues<this>): void {
```
Comment (zsarnett) [inline src/components/media-player/ha-media-player-browse.ts:150]: I just moved this. Nothing changed

File: src/components/media-player/ha-media-player-browse.ts [hunk 2]
```diff
@@ -25,7 +26,6 @@ import { styleMap } from "lit/directives/style-map";
 import { fireEvent } from "../../common/dom/fire_event";
 import { computeRTLDirection } from "../../common/util/compute_rtl";
 import { debounce } from "../../common/util/debounce";
-import { getSignedPath } from "../../data/auth";
```
Comment (balloob) [inline src/components/media-player/ha-media-player-browse.ts:28]: If this is no longer needed it means that thumbnails are broken 🙃

File: src/components/media-player/ha-media-player-browse.ts [hunk 3]
```diff
@@ -296,89 +436,38 @@ export class HaMediaPlayerBrowse extends LitElement {
                   `
                 : childrenMediaClass.layout === "grid"
                 ? html`
-                    <div
+                    <lit-virtualizer
```
Comment (bramkragten) [inline src/components/media-player/ha-media-player-browse.ts:439]: Missing import?

File: src/components/media-player/ha-media-player-browse.ts [hunk 4]
```diff
@@ -454,144 +543,61 @@ export class HaMediaPlayerBrowse extends LitElement {
...
+            ${child.title}
+            <paper-tooltip fitToVisibleBounds position="top" offset="4"
+              >${child.title}</paper-tooltip
+            >
+          </div>
+        </ha-card>
+      </div>
+    `;
 
   private _actionClicked(ev: MouseEvent): void {
```
Comment (bramkragten) [inline src/components/media-player/ha-media-player-browse.ts:602]:
```suggestion
  private _actionClicked = (ev: MouseEvent): void => {
```

File: build-scripts/webpack.js [hunk 5]
```diff
@@ -138,6 +138,7 @@ const createWebpackConfig = ({
         "lit/directives/cache$": "lit/directives/cache.js",
         "lit/directives/repeat$": "lit/directives/repeat.js",
         "lit/polyfill-support$": "lit/polyfill-support.js",
+        "@lit-labs/virtualizer/layouts/grid": "lit/polyfill-support.js",
```
Comment (bramkragten) [inline build-scripts/webpack.js:141]: Uuuhhhh this is not good 😄

Comment (bramkragten) [inline build-scripts/webpack.js:141]:
```suggestion
        "@lit-labs/virtualizer/layouts/grid": "@lit-labs/virtualizer/layouts/grid.js",
```

File: src/components/media-player/ha-media-player-browse.ts [hunk 6]
```diff
@@ -101,8 +101,6 @@ export class HaMediaPlayerBrowse extends LitElement {
 
   @query(".content") private _content?: HTMLDivElement;
 
-  @queryAll(".lazythumbnail") private _thumbnails?: HaCard[];
```
Comment (bramkragten) [inline src/components/media-player/ha-media-player-browse.ts:104]: Also remove the class from the elements, and remove `data-src`

Comment (zsarnett) [reply src/components/media-player/ha-media-player-browse.ts:104]: oh I guess this may still be used... for lists

Comment (bramkragten) [reply src/components/media-player/ha-media-player-browse.ts:104]: `_intersectionObserver` not used anymore?

Comment (bramkragten) [reply src/components/media-player/ha-media-player-browse.ts:104]: Or just virtualize the list too, it is easy :-)

File: src/components/media-player/ha-media-player-browse.ts [hunk 7]
```diff
@@ -173,81 +309,83 @@ export class HaMediaPlayerBrowse extends LitElement {
...
                             ? html`
-                                <mwc-button
-                                  raised
-                                  .item=${currentItem}
-                                  @click=${this._actionClicked}
+                                <div
+                                  class="img"
+                                  style=${styleMap({
+                                    backgroundImage: currentItem.thumbnail
+                                      ? `url(${currentItem.thumbnail})`
```
Comment (balloob) [inline src/components/media-player/ha-media-player-browse.ts:327]: This doesn't work. It needs to be signed. That's what the intersection observer was doing.

Comment (balloob) [reply src/components/media-player/ha-media-player-browse.ts:327]: It was also rewriting the brand pictures to support dark mode.

Comment (zsarnett) [reply src/components/media-player/ha-media-player-browse.ts:327]: fixing in next push

**Review by bramkragten — APPROVED** [review]

**Pull request merged**